src/safeposix/syscalls/fs_calls.rs

```diff
@@ -910,6 +909,421 @@ impl Cage {
+
+                    match inodeobj {
+                        Inode::File(ref mut normalfile_inode_obj) => {
+                            normalfile_inode_obj.refcount -= 1;
+
+                            //if there are still references to the file, then do nothing to the file
+                            if normalfile_inode_obj.refcount != 0 {
+                                return 0;
+                            }
+                            if fobjtable.contains_key(&inodenum) {
+                                //close and delete the file in the object table
+                                fobjtable.get(&inodenum).unwrap().close().unwrap();
+                                fobjtable.remove(&inodenum);
+                            }
+                            if normalfile_inode_obj.refcount == 0 {
+                                if normalfile_inode_obj.linkcount == 0  {
+                                    
+                                    //removing the file from the entire filesystem
+                                    mutmetadata.inodetable.remove(&inodenum);
+                                    let sysfilename = format!("{}{}", FILEDATAPREFIX, inodenum);
+                                    interface::removefile(sysfilename).unwrap();
+            
+                                } 
+                            }
+                        },
+                        Inode::Dir(ref mut dir_inode_obj) => {
+                            dir_inode_obj.refcount -= 1;
+
+                            //if there are still references to the file, then do nothing to the file
+                            if dir_inode_obj.refcount != 0 {
+                                return 0;
+                            }
+                            if fobjtable.contains_key(&inodenum) {
+                                //close and delete the file in the object table
+                                fobjtable.get(&inodenum).unwrap().close().unwrap();
+                                fobjtable.remove(&inodenum);
+                            }
+                            if dir_inode_obj.refcount == 0 {
+                                if dir_inode_obj.linkcount == 0  {
+                                    mutmetadata.inodetable.remove(&inodenum);
+                                    let sysfilename = format!("{}{}", FILEDATAPREFIX, inodenum);
+                                    interface::removefile(sysfilename).unwrap();
+            
+                                } 
+                            }
+                        },
+                        Inode::CharDev(ref mut char_inode_obj) => {
+                            char_inode_obj.refcount -= 1;
+
+                            //if there are still references to the file, then do nothing to the file
+                            if char_inode_obj.refcount != 0 {
+                                return 0;
+                            }
```

Comment:
You still want to go through with removing the file descriptor from the fdtable even if the refcount is non-zero, the code will need to be refactored to allow that